### PR TITLE
Check additional Docker config file paths ($HOME, .dockerconfigjson, and legacy .dockercfg)

### LINKED ISSUE #2260: Try .docker/.dockercfg and $HOME/.docker/... too
(From https://github.com/GoogleContainerTools/jib/issues/2258#issuecomment-580823413.)

It may (or may not) make sense to check `.dockercfg`. Looks like it was an [old config name](https://stackoverflow.com/a/35547703/1701388). Note `.dockercfg` has a [different JSON structure](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/private-auth-container-instances.html). Seems like `.dockercfg` is from a really old Docker version.

But more importantly, I think we should also try `$HOME` in addition to `System.getProperty("user.home")`: https://github.com/GoogleContainerTools/jib/issues/2260#issuecomment-580847849

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #2260.

In addition to the current

* System.get("user.home")/.docker/config.json

this PR now searches for

* System.get("user.home")/.docker/.dockerconfigjson
* System.get("user.home")/.docker/.dockercfg (legacy format)
* $HOME/.docker/config.json
* $HOME/.docker/.dockerconfigjson
* $HOME/.docker/.dockercfg (legacy format)

`$HOME` can be different from `System.get("user.home")`, which is the case on GCB. For example, after this fix, we will be able to eventually remove passing `-Duser.home=/home/builder` in Skaffold (although recent Jib versions can populate Google ADC and thus will work without `-Duser.home=` on GCB).

I learned that the legacy format was just the `auths` section of the new format. For example, suppose this is a new format file:

```
{
  "auths":{
    "some other registry":{"auth":"c29tZTpvdGhlcjphdXRo"},
    "some registry":{"auth":"c29tZTphdXRo","password":"ignored"},
    "https://registry":{"auth":"dG9rZW4="},

    "example.com":{"auth":"blah blah"}
  },
  "credsStore": ...,
  "credHelpers": ...
}
```

An example legacy file can be

```
{
  "some other registry":{"auth":"c29tZTpvdGhlcjphdXRo"},
  "some registry":{"auth":"c29tZTphdXRo","password":"ignored"},
  "https://registry":{"auth":"dG9rZW4="},

  "example.com":{"auth":"blah blah"}
}
```

### CODE REVIEW

**Comment by cmoulliard** [comment]
LGTM. I'm doing a test where the JIB build takes place within a pod running on ocp3 and the snapshot created from this PR is generating this error 
```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.0.1-SNAPSHOT:build (default-cli) on project quarkus-rest: Obtaining project build output files failed; make sure you have packaged your project before trying to build the image. (Did you accidentally run "mvn clean jib:build" instead of "mvn clean package jib:build"?): /home/jboss/quarkus-demo/target/quarkus-rest-1.0-SNAPSHOT.jar -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
sh-4.4$ ls -la target/total 8996drwxr-xr-x. 8 1000230000 root    4096 Feb  4 07:52 .
drwxr-xr-x. 8 1000230000 root    4096 Feb  4 07:52 ..
drwxr-xr-x. 4 1000230000 root    4096 Feb  4 07:52 classes
drwxr-xr-x. 3 1000230000 root    4096 Feb  4 07:52 generated-sources
drwxr-xr-x. 3 1000230000 root    4096 Feb  4 07:52 generated-test-sources
drwxr-xr-x. 2 1000230000 root    4096 Feb  4 07:52 maven-archiver
drwxr-xr-x. 3 1000230000 root    4096 Feb  4 07:52 maven-status
-rw-r--r--. 1 1000230000 root 9162851 Feb  4 07:52 quarkus-rest-1.0-SNAPSHOT-runner.jar
-rw-r--r--. 1 1000230000 root    5627 Feb  4 07:52 quarkus-rest-1.0-SNAPSHOT.jar.original
drwxr-xr-x. 3 1000230000 root    4096 Feb  4 07:52 test-classes
sh-4.4$ mvn clean package com.google.cloud.tools:jib-maven-plugin:2.0.1-SNAPSHOT:build -Dmaven.local.repo=/home/jboss/.m2/repository -Duser.home=/home/jboss -DsendCredentialsOverHttp=true -Djib.allowInsecureRegistries=true -Dimage=172.30.1.1:5000/test/quarkus-demo -Djib.from.image=172.30.1.1:5000/test/ubi11 -Djib.container.mainClass=io.quarkus.runner.GeneratedMain -Djib.containerizingMode=packaged -Dquarkus.package.uber-jar=true -DskipTests=true
```

**Comment by cmoulliard** [comment]
Test is working fine using old docker config 
```
[INFO] Using base image with digest: sha256:2b2e36eda2458df0d46681c4ee6273ca2cf1ffaebd15673be1a9952462356392
[INFO]
[INFO] Container entrypoint set to [java, -cp, /app/resources:/app/classes:/app/libs/*, io.quarkus.runner.GeneratedMain]
[INFO]
[INFO] Built and pushed image as 172.30.1.1:5000/test/quarkus-demo
[INFO] Executing tasks:
[INFO] [===========================   ] 88.9% complete
[INFO] > launching layer pushers
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  23.166 s
[INFO] Finished at: 2020-02-04T08:00:09Z
[INFO] ------------------------------------------------------------------------
sh-4.4$ ls -la ../.docker/.
total 8
drwxrwsrwt. 3 root  1000230000  100 Feb  4 07:46 .
drwxrwx---. 1 jboss root       4096 Feb  4 08:00 ..
drwxr-sr-x. 2 root  1000230000   60 Feb  4 07:46 ..2020_02_04_07_46_09.193927296
lrwxrwxrwx. 1 root  root         31 Feb  4 07:46 ..data -> ..2020_02_04_07_46_09.193927296
lrwxrwxrwx. 1 root  root         17 Feb  4 07:46 .dockercfg -> ..data/.dockercfg
```

**Comment by chanseokoh** [comment]
@cmoulliard it's currently not possible to configure the JAR path. In https://github.com/GoogleContainerTools/jib/issues/1555#issuecomment-562836425, @gunnarmorling used a trick of [using `copy-rename-maven-plugin`](https://github.com/gunnarmorling/quarkus-with-jib/blob/65f0ae7958b1754ec18ffeef89260bfe0ba3e362/pom.xml#L102-L119) to copy `quarkus-rest-1.0-SNAPSHOT-runner.jar` into `quarkus-rest-1.0-SNAPSHOT.jar`. The [README](https://github.com/gunnarmorling/quarkus-with-jib) of his demo repo says

> Jib will take the main artifact produced by the build and copy that into the image. As Quarkus needs its runner JAR to be copied instead, the copy-rename-maven-plugin is used to rename the runner JAR into the expected name.

And FYI again, https://github.com/GoogleContainerTools/jib/issues/1555#issuecomment-581492637

**Comment by cmoulliard** [comment]
> not possible to configure the JAR path

I tested successfully using copy plugin as suggested by Gunnar. We can go ahead for me and merge the PR @chanseokoh

**Comment by cmoulliard** [comment]
@chanseokoh I re-build this morning JIB snapshot top of this PR (commit : dfc960) and the test was again successful :-)

```
sh-4.4$ mvn clean package com.google.cloud.tools:jib-maven-plugin:2.0.1-SNAPSHOT:build -Dmaven.local.repo=/home/jboss/.m2/repository -Duser.home=/home/jboss -DsendCredentialsOverHttp=true -Djib.allowInsecureRegistries=true -Dimage=172.30.1.1:5000/test/quarkus-demo -Djib.from.image=172.30.1.1:5000/test/ubi11 -Djib.container.mainClass=io.quarkus.runner.GeneratedMain
[INFO] Scanning for projects...
[INFO]
[INFO] ---------------------< dev.snowdrop:quarkus-rest >----------------------
[INFO] Building quarkus-rest 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ quarkus-rest ---
[INFO] Deleting /home/jboss/quarkus-demo/target
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ quarkus-rest ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 3 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ quarkus-rest ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /home/jboss/quarkus-demo/target/classes
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ quarkus-rest ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/jboss/quarkus-demo/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ quarkus-rest ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /home/jboss/quarkus-demo/target/test-classes
[INFO]
[INFO] --- maven-surefire-plugin:2.22.1:test (default-test) @ quarkus-rest ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running dev.snowdrop.HelloApplicationTest
2020-02-07 08:33:12,277 INFO  [io.quarkus] (main) Quarkus 1.2.0.Final started in 1.134s. Listening on: http://0.0.0.0:8081
2020-02-07 08:33:12,280 INFO  [io.quarkus] (main) Profile test activated.
2020-02-07 08:33:12,280 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 10.779 s - in dev.snowdrop.HelloApplicationTest
2020-02-07 08:33:15,302 INFO  [io.quarkus] (main) Quarkus stopped in 0.134s
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ quarkus-rest ---
[INFO] Building jar: /home/jboss/quarkus-demo/target/quarkus-rest-1.0-SNAPSHOT.jar
[INFO]
[INFO] --- quarkus-maven-plugin:1.2.0.Final:build (default) @ quarkus-rest ---
[INFO] [org.jboss.threads] JBoss Threads version 3.0.0.Final
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building thin jar: /home/jboss/quarkus-demo/target/quarkus-rest-1.0-SNAPSHOT-runner.jar
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 2597ms
[INFO]
[INFO] --- copy-rename-maven-plugin:1.0.1:rename (rename-runner) @ quarkus-rest ---
[INFO] Renamed /home/jboss/quarkus-demo/target/quarkus-rest-1.0-SNAPSHOT-runner.jar to /home/jboss/quarkus-demo/target/quarkus-rest-1.0-SNAPSHOT.jar
[INFO]
[INFO] --- jib-maven-plugin:2.0.1-SNAPSHOT:build (default-cli) @ quarkus-rest ---
[WARNING] Authentication over HTTP is enabled. It is strongly recommended that you do not enable this on a public network!
[INFO]
[INFO] Containerizing application to 172.30.1.1:5000/test/quarkus-demo...
[WARNING] Failed to read global Jib config: No content to map due to end-of-input
 at [Source: (sun.nio.ch.ChannelInputStream); line: 1, column: 0]; you may need to fix or delete /home/jboss/.config/google-cloud-tools-java/jib/config.json;
[WARNING] Base image '172.30.1.1:5000/test/ubi11' does not use a specific image digest - build may not be reproducible
[INFO] Using credentials from Docker config (/home/jboss/.docker/.dockercfg) for 172.30.1.1:5000/test/quarkus-demo
[WARNING] Cannot verify server at https://172.30.1.1:5000/v2/. Attempting again with no TLS verification.
[WARNING] Cannot verify server at https://172.30.1.1:5000/v2/test/ubi11/manifests/latest. Attempting again with no TLS verification.
[WARNING] Failed to connect to https://172.30.1.1:5000/v2/test/ubi11/manifests/latest over HTTPS. Attempting again with HTTP.
[WARNING] Failed to connect to https://172.30.1.1:5000/v2/ over HTTPS. Attempting again with HTTP.
[INFO] The base image requires auth. Trying again for 172.30.1.1:5000/test/ubi11...
[INFO] Using credentials from Docker config (/home/jboss/.docker/.dockercfg) for 172.30.1.1:5000/test/ubi11
[WARNING] Cannot verify server at https://172.30.1.1:5000/v2/test/ubi11/manifests/latest. Attempting again with no TLS verification.
[WARNING] Failed to connect to https://172.30.1.1:5000/v2/test/ubi11/manifests/latest over HTTPS. Attempting again with HTTP.
[WARNING] Cannot verify server at https://172.30.1.1:5000/v2/test/quarkus-demo/blobs/sha256:07f05d9b711882bc653b23f62b570916cbbb56a36d8243ab02b938c7eadaeea8. Attempting again with no TLS verification.
[WARNING] Cannot verify server at https://172.30.1.1:5000/v2/test/quarkus-demo/blobs/sha256:7bb096a9c8741bf8e8a2ac2854f69655dd9df5d9eed014efdeff172334e209de. Attempting again with no TLS verification.
[WARNING] Failed to connect to https://172.30.1.1:5000/v2/test/quarkus-demo/blobs/sha256:7bb096a9c8741bf8e8a2ac2854f69655dd9df5d9eed014efdeff172334e209de over HTTPS. Attempting again with HTTP.
[INFO] Using base image with digest: sha256:2b2e36eda2458df0d46681c4ee6273ca2cf1ffaebd15673be1a9952462356392
[WARNING] Failed to connect to https://172.30.1.1:5000/v2/test/quarkus-demo/blobs/sha256:07f05d9b711882bc653b23f62b570916cbbb56a36d8243ab02b938c7eadaeea8 over HTTPS. Attempting again with HTTP.
[INFO]
[INFO] Container entrypoint set to [java, -cp, /app/resources:/app/classes:/app/libs/*, io.quarkus.runner.GeneratedMain]
[INFO]
[INFO] Built and pushed image as 172.30.1.1:5000/test/quarkus-demo
[INFO] Executing tasks:
[INFO] [===========================   ] 88.9% complete
[INFO] > launching layer pushers
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  30.922 s
[INFO] Finished at: 2020-02-07T08:33:23Z
[INFO] ------------------------------------------------------------------------
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java [hunk 1]
```diff
@@ -187,17 +187,6 @@ public CredentialRetriever wellKnownCredentialHelpers() {
...
   }
 
-  /**
-   * Creates a new {@link CredentialRetriever} that tries to retrieve credentials from Docker config
-   * (located at {@code $USER_HOME/.docker/config.json}).
-   *
-   * @return a new {@link CredentialRetriever}
-   * @see DockerConfigCredentialRetriever
-   */
-  public CredentialRetriever dockerConfig() {
```
[the remark this answers is not in the document]
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java]: So do we look in `$USER_HOME` first and if no file exists then check `user.home`?

[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java]: As of now, we only check `user.home`. With this PR, we first check `user.home`, and if we fail, we try `$HOME`. We don't introduce a breaking change.

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerConfigCredentialRetriever.java [hunk 2]
```diff
@@ -81,6 +85,20 @@ public Path getDockerConfigFile() {
     if (!Files.exists(dockerConfigFile)) {
       return Optional.empty();
     }
+
+    if (legacyConfigFormat) {
```
Comment (loosebazooka) [inline jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerConfigCredentialRetriever.java]: potential value in adding comment here (or somewhere) that `legacyConfigFormat` = `auths{}` block of dockerConfig.

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerConfigCredentialRetriever.java]: I was able to simplify this a bit.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (briandealwis) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java]: I must be missing something but from what I can see you've removed this arg-less function and replaced with with two other variants that require a pointer to a file.  We use this method in the [jib-core examples README](https://github.com/GoogleContainerTools/jib/blob/master/jib-core/examples/build.gradle/README.md)

[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java]: I was talking to @loosebazooka. The method is restored.

**Comment by chanseokoh** [comment]
> I think I'm okay with this, would check with @briandealwis since he had some comments earlier

This is resolved. Going to merge it. But I can always fix something after that.

**Pull request merged**